nexus/client.go

```diff
@@ -0,0 +1,629 @@
+		Operation: operation,
+		ID:        operationID,
+	}, nil
+}
+
+func (c *Client) sendGetOperationResultRequest(ctx context.Context, httpReq *http.Request, wait time.Duration) (*http.Response, error) {
+	if wait > 0 {
+		timeout := min(wait, c.Options.GetResultMaxTimeout)
+		if deadline, set := ctx.Deadline(); set {
+			timeout = min(timeout, time.Until(deadline))
+		}
+
+		url := httpReq.URL
+		q := url.Query()
+		q.Set(queryWait, fmt.Sprintf("%dms", timeout.Milliseconds()))
+		url.RawQuery = q.Encode()
+	}
+
+	response, err := c.Options.HTTPCaller(httpReq)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode == http.StatusOK || response.StatusCode == http.StatusNoContent {
+		return response, nil
+	}
+
+	// Do this once here and make sure it doesn't leak.
+	body, err := readAndReplaceBody(response)
+	if err != nil {
+		return nil, err
+	}
+
+	switch response.StatusCode {
+	case statusOperationRunning:
+		return nil, ErrOperationStillRunning
+	case statusOperationFailed:
+		state, err := getUnsuccessfulStateFromHeader(response, body)
+		if err != nil {
+			return nil, err
+		}
+		failure, err := failureFromResponse(response, body)
+		if err != nil {
+			return nil, err
+		}
+		return nil, &UnsuccessfulOperationError{
+			State:   state,
+			Failure: failure,
+		}
+	default:
+		return nil, newUnexpectedResponseError(fmt.Sprintf("unexpected response status: %q", response.Status), response, body)
+	}
+}
+
+// DeliverCompletion delivers the result of a completed asynchronous operation to the provided URL.
+// If completion is an [OperationCompletionSuccessful] its body will be automatically closed.
+func (c *Client) DeliverCompletion(ctx context.Context, url string, completion OperationCompletion) error {
+	// while the http client is expected to close the body, we close in case request creation fails (which may double close but that's fine since we ignore the error).
+	defer completion.Close()
+	httpReq, err := http.NewRequestWithContext(ctx, "POST", url, nil)
+	if err != nil {
+		return err
+	}
+	if err := completion.applyToHTTPRequest(httpReq, c); err != nil {
+		return err
+	}
+
+	httpReq.Header.Set(headerUserAgent, userAgent)
+	response, err := c.Options.HTTPCaller(httpReq)
+	if err != nil {
+		return err
+	}
+
+	// Do this once here and make sure it doesn't leak.
+	body, err := readAndReplaceBody(response)
+	if err != nil {
+		return err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return newUnexpectedResponseError(fmt.Sprintf("unexpected response status: %q", response.Status), response, body)
+	}
+
+	return nil
+}
+
+func (c *Client) joinURL(parts ...string) (*url.URL, error) {
+	if c.serviceBaseURL == nil {
+		return nil, errEmptyServiceBaseURL
+	}
+	return c.serviceBaseURL.JoinPath(parts...), nil
+}
+
+// OperationCompletion is input for [Client.DeliverCompletion].
+// It has two implementations: [OperationCompletionSuccessful] and [OperationCompletionUnsuccessful].
+type OperationCompletion interface {
+	io.Closer
```

Comment:
Yeah, I was debating if that was important or not.
Figured io.Closer was quite standard.

Reply:
It's not whether it's a standard, it's about API surface area. You should not infect the user-facing API surface with an internal-only call. If the caller needed to close this instead of this code, I would agree, you expose `Close()`.

Reply:
They may want to close it, if they create this completion and hit an error before passing it over to `DeliverCompletion`.

Reply:
They have the body to close if they need. We will close it only if it gets returned to us. But we don't need to give them a helper to close their own thing. It's not valuable API area.